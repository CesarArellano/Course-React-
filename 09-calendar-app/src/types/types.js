
export const types = {
  uiOpenModal: '[UI] Open modal',
  uiCloseModal: '[UI] Close modal',

  eventSetActive: '[EVENT] Set active',
  eventAddNew: '[EVENT] Add new',
  eventClearActiveEvent: '[EVENT] Clear active event',
  eventUpdate: '[EVENT] Event updated'
};  