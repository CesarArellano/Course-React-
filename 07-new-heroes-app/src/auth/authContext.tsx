import { createContext } from "react";

export interface AuthState {
  user: {
    name: string;
    logged: boolean;
  }
  dispatch: React.Dispatch<any>
}

export const AuthContext = createContext({} as AuthState );