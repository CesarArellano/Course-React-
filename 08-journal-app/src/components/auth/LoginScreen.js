import React from 'react'
import { Link } from 'react-router-dom'

export const LoginScreen = () => {
  return (
    <>
      <div className="center-box">
        <img 
          src="https://image.flaticon.com/icons/png/512/3235/3235042.png"
          alt="Journal App Icon" 
          title="App Icon"
          className="img-icon"
        />
      </div>
      <h3 className="auth__title">Login</h3>
      <form>
        <input
          type="text"
          placeholder="Email"
          name="email"
          autoComplete="off"
          className="auth__input"
        />

        <input 
          type="password"
          placeholder="Password"
          name="password"
          className="auth__input"
        />

        <div className="center-box">
          <button
            type="submit"
            className="btn btn-primary mb-5"
          >
            Login
          </button>
        </div>
        <hr />
        <div className="auth__social-networks">
          <p>Login with social networks</p>
          <div className="google-btn">
              <div className="google-icon-wrapper">
                  <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
              </div>
              <p className="btn-text">
                  <b>Sign in with google</b>
              </p>
          </div>
        </div>
          <Link 
            to="/auth/register"
            className="link"
          >
            Create New Account
          </Link>
      </form>
    </>
  )
}
