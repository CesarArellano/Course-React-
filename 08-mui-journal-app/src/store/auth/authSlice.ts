import { createSlice } from '@reduxjs/toolkit';

export interface AuthReducerProps {
  status: 'checking' | 'not-authenticated' | 'authenticated';
  uid: string | null;
  email: string | null;
  displayName: string | null;
  photoURL: string | null;
  errorMessage: string | null;
}

export const authSlice = createSlice({
  name: 'auth',
  initialState: {
    status: 'not-authenticated', // 'not-authenticated', 'authenticated'
    uid: null,
    email: null,
    displayName: null,
    photoURL: null,
    errorMessage: null,
  } as AuthReducerProps,
  reducers: {
    login: ( state, action) => {
      
    },
    logout: ( state, action) => {
      
    },
    checkingCredentials: ( state, action) => {
      state.status = 'checking';
    },
  }
});


export const { login, logout, checkingCredentials } = authSlice.actions;