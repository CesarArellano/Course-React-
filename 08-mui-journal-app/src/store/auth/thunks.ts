import { AnyAction, Dispatch } from '@reduxjs/toolkit';
import { checkingCredentials } from './';

export const checkingAuthentication = (email: string, password: string) => {
  return async( dispatch: Dispatch<AnyAction>) => {
    dispatch( checkingCredentials('') )
  }
}

export const startGoogleSignIn = () => {
  return async( dispatch: Dispatch<AnyAction>) => {
    dispatch( checkingCredentials('') )
  }
}