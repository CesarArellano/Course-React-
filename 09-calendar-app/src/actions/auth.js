import { fetchWithoutToken } from "../helpers/fetch";
import { types } from "../types/types";

export const startLogin = ( email, password ) => {
  return async ( dispatch ) => {
    const resp = await fetchWithoutToken('auth', { email, password }, 'POST');
    const decodedData = await resp.json();
    
    if( decodedData.ok ) {
      localStorage.setItem( 'token', decodedData.token );
      localStorage.setItem( 'token-init-date', new Date().getTime() );
      dispatch( login({
        uid: decodedData.uid,
        name: decodedData.name,
      }));
    }
  }
};

const login = ( user ) => ({
  type: types.authLogin,
  payload: user,
});