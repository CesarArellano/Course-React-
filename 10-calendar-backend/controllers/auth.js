const { response } = require('express');
const User = require('../models/User');

const createUser = async ( req, res = response ) => {
  
  const { email, password } = req.body;
  try {
    let user = await User.findOne({ email })

    if( user ) {
      return res.status(400).json({
        ok: false,
        msg: 'El usuario ya existe con ese correo.'
      });
    }

    user = new User( req.body );
    user.save();
    
    res.status(201).json({
      ok: true,
      uid: user.id,
      name: user.name
    });

  } catch(error) {
    console.log(error);
    res.status(500).json({
      ok: false,
      msg: 'Please talk to the admin'
    });
  }
  
}

const userLogin = ( req, res = response ) => {
  const { email, password } = req.body;

  res.json({
    ok: true,
    msg: 'login',
    email,
    password
  });

};

const revalidateToken = ( req, res = response ) => {
  res.json({
    ok: true,
    msg: 'renew'
  });
};

module.exports = {
  createUser,
  userLogin,
  revalidateToken
};