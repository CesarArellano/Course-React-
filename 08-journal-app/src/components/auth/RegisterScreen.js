import React from 'react'
import { Link } from 'react-router-dom'
import { useDispatch, useSelector } from 'react-redux';
import { useForm } from '../../hooks/useForm'
import validator from 'validator';
import { removeError, setError } from '../../actions/ui';
import { startRegisterWithEmailPasswordName } from '../../actions/auth';

export const RegisterScreen = () => {
  const dispatch = useDispatch();
  const { ui:{ msgError } } = useSelector( state => state); // Recover state data
  
  const [ formValues, handleInputChange ] = useForm({
    name: 'CÃ©sar',
    email: 'cesarmauricio.arellano@gmail.com',
    password: '12345678',
    password2: '12345678'
  });

  const { name, email, password, password2 } = formValues;
  
  const handleRegister = (e) => {
    e.preventDefault();
    
    if( isFormValid() ) {
      dispatch( startRegisterWithEmailPasswordName( email, password, name) )
    }
  };

  const isFormValid = () => {
    if( name.trim().length === 0 ) {
      dispatch(setError('Name is required'));
      return false;
    } else if( !validator.isEmail(email) ){
      dispatch(setError('Email is not valid'));
      return false;
    } else if ( password !== password2 || password.length < 8 ) {
      dispatch(setError('Password should be at least 8 characters and match each other'));
      return false;
    }
    dispatch( removeError() );
    return true;
  };

  return (
    <>
      <div className="center-box">
        <img 
          src="https://image.flaticon.com/icons/png/512/3235/3235042.png"
          alt="Journal App Icon" 
          title="App Icon"
          className="img-icon"
        />
      </div>
      <h3 className="auth__title">Register</h3>
      <form onSubmit={ handleRegister }>
        {
          ( msgError ) && <div className="auth__alert-error animate__animated animate__shakeX">
            { msgError }
          </div>
        }
        
        <input
          type="text"
          placeholder="Name"
          name="name"
          autoComplete="off"
          className="auth__input"
          value={ name }
          onChange={ handleInputChange }
          required
        />
        <input
          type="text"
          placeholder="Email"
          name="email"
          autoComplete="off"
          className="auth__input"
          value={ email }
          onChange={ handleInputChange }
        />

        <input 
          type="password"
          placeholder="Password"
          name="password"
          className="auth__input"
          value={ password }
          onChange={ handleInputChange }
        />

        <input 
          type="password"
          placeholder="Confirm password"
          name="password2"
          className="auth__input"
          value={ password2 }
          onChange={ handleInputChange }
        />

        <div className="center-box">
          <button
            type="submit"
            className="btn btn-primary mb-5"
          >
            Register
          </button>
        </div>
          <Link 
            to="/auth/login"
            className="link"
          >
            Already registered?
          </Link>
      </form>
    </>
  )
}
