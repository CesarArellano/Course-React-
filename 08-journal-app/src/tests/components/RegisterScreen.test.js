import '@testing-library/jest-dom';

import { mount } from "enzyme"
import { RegisterScreen } from "../../components/auth/RegisterScreen";
import { Provider } from "react-redux";
import thunk from "redux-thunk";
import configureStore from "redux-mock-store";
import { MemoryRouter } from "react-router-dom";
import { types } from '../../types/types';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initState = {
  ui: {
    msgError: 'Email is not valid',
    loading: false
  }
};

const store = mockStore(initState);

const wrapper = mount(
  <Provider store={ store }>
    <MemoryRouter>
      <RegisterScreen />
    </MemoryRouter>
  </Provider>
);

describe('Testing in <RegisterScreen />', () => {
  
  test('should show the component correctly', () => {
    
    expect(wrapper).toMatchSnapshot();
  });

  test('should show alert', () => {
    const emailField = wrapper.find('input[name="email"]');
    emailField.simulate('change', {
      target: {
        value: '',
        name: 'email',
      }
    });
    
    wrapper.find('form').simulate('submit', { 
      preventDefault(){} 
    });

    const actions = store.getActions();

    expect(actions[0]).toEqual({
      type: types.uiSetError,
      payload: 'Email is not valid'
    });
  })
  
  test('should show the alert box with the msg', () => {
    expect( wrapper.find('.auth__alert-error').exists() ).toBe( true );
    expect( wrapper.find('.auth__alert-error').text().trim() ).toBe( initState.ui.msgError );
  })
  
  
})
