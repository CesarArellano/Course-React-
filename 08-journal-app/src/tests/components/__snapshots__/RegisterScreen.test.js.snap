// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing in <RegisterScreen /> should show the component correctly 1`] = `
Array [
  <div
    className="center-box"
  >
    <img
      alt="Journal App Icon"
      className="img-icon"
      src="https://image.flaticon.com/icons/png/512/3235/3235042.png"
      title="App Icon"
    />
  </div>,
  <h3
    className="auth__title"
  >
    Register
  </h3>,
  <form
    onSubmit={[Function]}
  >
    <div
      className="auth__alert-error animate__animated animate__shakeX"
    >
      Email is not valid
    </div>
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      required={true}
      type="text"
      value="CÃ©sar"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value="cesarmauricio.arellano@gmail.com"
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value="12345678"
    />
    <input
      className="auth__input"
      name="password2"
      onChange={[Function]}
      placeholder="Confirm password"
      type="password"
      value="12345678"
    />
    <div
      className="center-box"
    >
      <button
        className="btn btn-primary mb-5"
        type="submit"
      >
        Register
      </button>
    </div>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      Already registered?
    </a>
  </form>,
]
`;
